server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -8,32 +8,194 @@
+    private final ThreadPool threadPool;
+    private final ClusterService clusterService;
+    private final WorkloadManagementSettings workloadManagementSettings;
+    private final Set<QueryGroup> activeQueryGroups;
+    private final Set<QueryGroup> deletedQueryGroups;
+    private final NodeDuressTrackers nodeDuressTrackers;
 
-    public QueryGroupService() {
-        this(new HashMap<>());
+    public QueryGroupService(
+        QueryGroupTaskCancellationService taskCancellationService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings
+    ) {
+
+        this(
+            taskCancellationService,
+            clusterService,
+            threadPool,
+            workloadManagementSettings,
+            new NodeDuressTrackers(
+                Map.of(
+                    ResourceType.CPU,
+                    new NodeDuressTracker(
+                        () -> workloadManagementSettings.getNodeLevelCpuCancellationThreshold() < ProcessProbe.getInstance()
+                            .getProcessCpuPercent() / 100.0,
+                        workloadManagementSettings::getDuressStreak
+                    ),
+                    ResourceType.MEMORY,
+                    new NodeDuressTracker(
+                        () -> workloadManagementSettings.getNodeLevelMemoryCancellationThreshold() <= JvmStats.jvmStats()
+                            .getMem()
+                            .getHeapUsedPercent() / 100.0,
+                        workloadManagementSettings::getDuressStreak
+                    )
+                )
+            ),
+            new HashMap<>(),
+            new HashSet<>(),
+            new HashSet<>()
+        );
     }
 
-    public QueryGroupService(Map<String, QueryGroupState> queryGroupStateMap) {
-        this.queryGroupStateMap = queryGroupStateMap;
+    public QueryGroupService(
+        QueryGroupTaskCancellationService taskCancellationService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        WorkloadManagementSettings workloadManagementSettings,
+        NodeDuressTrackers nodeDuressTrackers,
+        Map<String, QueryGroupState> stateMap,
+        Set<QueryGroup> activeQueryGroups,
+        Set<QueryGroup> deletedQueryGroups
+    ) {
+        this.taskCancellationService = taskCancellationService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.nodeDuressTrackers = nodeDuressTrackers;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        activeQueryGroups.forEach(queryGroup -> stateMap.putIfAbsent(queryGroup.get_id(), new QueryGroupState()));
+        this.queryGroupStateMap = stateMap;
+        this.queryGroupStateMap.put(QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get(), new QueryGroupState());
```

Comment:
The cluster state at this point will not be initialised and if the metadata is reloaded then that would also be published to all nodes. In that case the active groups will be initialised at that point. 

Do you think otherwise ?

Reply:
No, that should work. Thanks!